### PR TITLE
Inline tests support

### LINKED ISSUE #65: Support for inline tests/benchmarks
Some thoughts on how to add automatic support for inline tests/benchmarks in jbuilder.

Currently [Jane Street jenga rules](https://github.com/janestreet/jenga-rules) support linking inline test and/or benchmark runners as soon as ones uses ppx_inline_test and/or ppx_bench. Additionally they automatically run the tests in parallel as part of the `runtest` alias.

This is a really useful feature for development and it would be great to add this to jbuilder. The rules are not very complicated to setup, however I'd like to support this without hardcoding the knowledge of the various libraries and their layout (ppx_bench, ppx_inline_tests, ppx_expect, ...) inside jbuilder. This is for two reason:
- to avoid having to update jbuilder whenever the details of these subsystems change. They don't change very often but do change from now and then
- to allow other people to reuse the same mechanism for their own test/benchmark systems

### The runners

For any given library Jbuilder systematically setups rules for building `<lib>-test-runner.exe` and `<lib>-bench-runner.exe` executables. These executables link all of library `<lib>` and are used for running tests and benchmarks respectively.

### Additional runner libraries

The various runners require additional libraries, essentially to report the results at the end. We need two things: a way to specify these additional libraries and a way to specify the code to run at the end of the runner. Currently the jenga rules hardcode this information and know what .cmx/.cmo to link last fo each runner.

#### Specifying additional libraries

We can simply add them in the jbuild of the ppx:

```scheme
(library
 ((name ppx_inline_test)
  ...
  (ppx_runner_library (test ppx_inline_test.runner))))
```

and Jbuilder will know to link `ppx_inline_test.runner` in the test runner of any libraries using this ppx.

#### Specifying the reporter

To make things simple, we provide a small `libmain` library allowing one to set the entry point:

```ocaml
(** Set the main entry point of the program *)
val set_main : (unit -> unit) -> unit

(** Execute the registered entry point. Never returns *)
val run_main : unit -> _

(** Are we linked as part of a library test/benchmark runner or a normal executable? *)
val is_library_runner : unit -> bool
```

and jbuilder will add a module that calls `run_main` as the last .cmx/.cmo on the command line to link the runner.

### running tests in parallel

We formalise the convention of ppx_inline_test for running tests in parallel: `foo-runner.exe --list-partitions` returns a list of partitions and `... --partition blah` selects tests from partition `blah`.

### Runtime test dependencies

Sometime inline tests require extra dependencies when running. For instance when they read a file. The jenga rules support this by looking for an `(inline_tests ...)` field in `(library ...)` stanzas:

```scheme
(library
 ((name foo)
  ...
  (inline_tests ((deps (blah))))))
```

We can do the same in jbuilder.

### helper scripts

The jenga rules generate helper shell scripts allowing one to run the tests/benchmarks manually. Instead, ppx_inline_test can provide an executable to call the generated runner appropriately. Alternatively we could extend `jbuilder exec` so that one can do:

```
$ jbuilder exec ./foo-test-runner.exe ...
```

### PR DESCRIPTION
**Pull request by rgrinberg**
- [x] Add `ppx_runner_library` to jbuild stanza
- [x] Add `inline_tests` to jbuild stanza
- [x] Setup rules for `inline_tests`
- [x] Generate meta file for test runner deps
- [x] Make `runtest` execute inline tests. While this is done, I'm not sure what the partitions stuff is all about in [this comment](https://github.com/janestreet/jbuilder/issues/65#issue-225402100) 
- [x] Add support for ppx_bench
- [x] Add bench alias
- [x] ~Properly implement libmain. `is_library_runner` isn't implemented currently.~ There's no need to implement libmain anymore.
- [x] ~Fix executing bytecode tests. Currently, those don't work because the stubs to enable tests don't get linked in.~ This is a general issue with bytecode and stubs.
- [x] Implement a proper run command for benchmarks.
- [ ] test benchmarks alias

Some notes:

* I've added the mutual recursion in gen_rules only to minimize moving around functions for the diff. I will get rid of it for the final commit.
* Again, I've found it easier to work with the higher level functions in gen_rules rather than low level ones like `build_exe`. I'm open to changing this however.
* I don't quite understand all the META generation and findlib compatibility stuff, so that stuff is likely wrong.
* `test-runner` doesn't work as a predicate name, i don't think dashes are allowed.
* We need to decide where libmain is going to live.
* I have a hacked version of ppx_inline_test that works with this branch. Will upload it in case anyone's interested.
* Is there any concern for ppx_jane/ppx_driver here? I haven't thought about how those might be relevant to this work.
 
Fix #65

### CODE REVIEW

**Comment by yminsky** [comment]
Any progress on this? I'm really eager to get this one moving.

**Comment by rgrinberg** [comment]
I think the essential questions still stand. Though the situation with libmain is likely resolved - it will live here in jbuilder. In the meantime, I've added a few relevant tests to this work (meta generation), I should also some simple tests with an example runner - that will also be useful as an example for documentation purposes. I'll rebase this work on master and work on the above for now.

Let's ping @diml, see if he can give a few pointers and unblock the important stuff.

**Comment by diml** [comment]
For libmain, what are the advantages of having at a separate package rather than having it as `jbuilder.libmain`? It seems to me that the latter will make our life easier.

> Again, I've found it easier to work with the higher level functions in gen_rules rather than low level ones like build_exe. I'm open to changing this however.

No objections here. However I'm a bit worried about implicitly generating a .ml file. Let's do the same as for ppx drivers and link a special library last. This special library will be linked with `-linkall` and will simply call `Libmain.run_main`. We can still use the high-level function, for instance you can extend the executables record to support that but not expose it to the user via the S-expression syntax.

> I don't quite understand all the META generation and findlib compatibility stuff, so that stuff is likely wrong.

At first glance it looks right, I'll need to read that properly again. BTW you could add some tests about this in the META generation test.

> test-runner doesn't work as a predicate name, i don't think dashes are allowed.

`test_runner` seems fine.

> I have a hacked version of ppx_inline_test that works with this branch. Will upload it in case anyone's interested.

Yes, please do

> Is there any concern for ppx_jane/ppx_driver here? I haven't thought about how those might be relevant to this work.

there shouldn't be

**Comment by rgrinberg** [comment]
> For libmain, what are the advantages of having at a separate package rather than having it as jbuilder.libmain? It seems to me that the latter will make our life easier.

I started it as `jbuilder.libmain` but leaning towards to `libmain` on its own for the following reasons:

* I don't like using the `x.y` findlib convention because I don't like relying on findlib at all and I don't like how dotted packages aren't discoverable. It's also kind of weird because we don't have a `jbuilder` toplevel package.
* I wanted to use the `(package libmain)` functionality to mark the libmain only tests as those need to run in-tree but still have external deps such as omp.
* Users who will have inline tests, will not be able to use `jbuilder` as a `{build}` dependency anymore. They will also need to remember it's a runtime deps for their tests. While if libmain is explicit they can just do `jbuilder {build}` and `libmain {test}`.

Though I don't think that these arguments are entirely convincing. I'm happy to revert to using `jbuilder.main` if you'd like.

**Comment by diml** [comment]
Ok, that makes sense to me. Happy to go for a separate package

**Comment by rgrinberg** [comment]
@dim OK so, I added a `libmain.main` library rather than generating a module. But unfortunately, `executable_rules` still requires a module to exist for a binary to be built. It's not so hard to work around this but it's but the way I've done it is pretty ugly. Will put up the commits that show the progression.

**Comment by dim** [comment]
@diml, Not @dim :)

**Comment by diml** [comment]
Ok, we can leave it like this for now. Later we can refactor a bit this code, we should split the interpretation of the jbuild stanza and the generation of the rules into separate functions, so that we have functions to produce executables that are easier to use grammatically.

**Comment by rgrinberg** [comment]
OK I've pushed some commits that implement preliminary support for ppx_bench. Things are now done the dumbest way possible, so will surely require refactoring. One question that is particular for ppx_bench support, what is the action to run the benchmarks?

**Comment by yminsky** [comment]
I'm curious, what's the plan for expect tests? It would be good to know what the intended workflow, so we can compare it to what we do internally. I think the rough story is that we generate a ".corrected" file for every failing test, and you can accept a new result by moving the corrected file on top of the original.

**Comment by rgrinberg** [comment]
> I'm curious, what's the plan for expect tests?

The plan is orthogonal to this PR. As you've said, .interacting with .corrected files is what's necessary to make expect tests work. So we plan to have a command to accept these .corrected files. It will be applicable for tests, linting, and also "inline" rules.

**Comment by diml** [comment]
To clarify the meaning of `is_library_runner`: `let%test ...` is translated by `ppx_inline_test` to a call of the form:

```ocaml
let () = Ppx_inline_test_lib.Runtime.register (fun () -> ...)
```

The register function will behave differently depending on whether the library is linked as part of a regular executable or a test runner. Currently we do that by exposing an external function [am_testing](https://github.com/janestreet/base/blob/master/src/import0.ml#L248) in Base. The C functions always return `false` and is marked as a weak definition. In the ppx_inlnie_test runner, we redifine the C function without the weak and make it always return `true`. This way, when we link a test runner, `Base.am_testing` is `true` while in all other cases it is `false`. For instance you can see uses of it in the [ppx_inline_test runtime](https://github.com/janestreet/ppx_inline_test/blob/master/runtime-lib/runtime.ml#L224).

`is_library_runner` is a generalization of this. Although, thinking about it again it might not be precise enough to distinguish between tests and benchmarks. We shouldn't need it initially either. Moreover, the better way to implement it is using library variants rather than C functions, since the former will work in all contexts, i.e. including js_of_ocaml.

So we can leave this feature out for now.

**Comment by rgrinberg** [comment]
@diml I've added a gross hack to make the tests work in isolation. Making the the expect test work outside the current framework proved to be a little difficult. Is this acceptable?

**Comment by rgrinberg** [comment]
Here's the ppx_inline_test branch https://github.com/janestreet/ppx_inline_test/pull/11/files

**Comment by rgrinberg** [comment]
Does it make sense for a library to be a ppx_runner_library for benchmarks and tests at the same time? Currently one must make choice. The utility for this probably tiny but the restriction seems a bit artificial as well.

I'm stumbling upon this when trying to reuse the same rewriter for tests/benchmarks. I can get around this problem easily via another means, but it's worth mentioning.

**Comment by diml** [comment]
I don't mind either way. We can change the syntax to be `(ppx_runner_libraries ((test ...) (bench ...)))`.

For the hack, I can't think of a better way indeed.

**Comment by rgrinberg** [comment]
Another thing that might be worth pointing getting right early: Should we generalize runner deps? Right now we have `inline_tests` but this doesn't really work for benchmarks. What about something like:

```
(runner_deps
 ((test (...))
  (bench (...))))
```

**Comment by yminsky** [comment]
I don't feel enormously strongly about this, but I do think that one of the nice things about the system we have internally is that inline tests just work, with no special build system work.

Another thought: if we don't support inline tests without a special stanza, it would quite a bit nicer if the inclusion of an inline test in the absence of this stanza led to a build error. Otherwise, people will run up against tests that they think are running, that aren't.

**Comment by rgrinberg** [comment]
If we add such stanzas, I wonder if it would be possible to automate inserting them with Leo's environment plugin. For situations where every library is expected to have such tests. There's already a bit of boilerplate in setting up tests for jbuilder, I'm not so keen on making it worse.

> Another thought: if we don't support inline tests without a special stanza, it would quite a bit nicer if the inclusion of an inline test in the absence of this stanza led to a build error. Otherwise, people will run up against tests that they think are running, that aren't.

Should be easy enough. Once the user adds ppx_inline_test to the list of preprocessors, we can check that if the preprocessor has runner dependencies, it should also include an associated runner stanza. On the other hand, it does make it strange that we're simply not doing this automatically for the user.

I'm not sure what the right thing to do here, but I don't really like making artificial differences between benchmarks and tests, and I agree that the concept of a runner is murky to the user.

**Comment by rgrinberg** [comment]
@diml note the last commit I just added. It was necessary to make sure the runners actually built before executing. Was this the intended behavior of add_alias, not to record this dependency?

**Comment by diml** [comment]
> @diml note the last commit I just added. It was necessary to make sure the runners acually built before executing. Was this the intended behavior of add_alias, not to record this dependency?

Ok. Why do you think it's related to `add_alias`? Basically if you specify an action manually, you have to specify the dependencies manually as well. I suppose we could use `Action.Infer.infer` systematically to avoid having to specify such dependencies manually. Currently we only do it for user actions. That might be worth investigating.

**Comment by rgrinberg** [comment]
> Ok. Why do you think it's related to add_alias?

I thought maybe there's a way to implement `add_alias` using `SC.Action.run` to do this tracking automatically. OK, so it was just a simple omission.

File: libmain/jbuild [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  (public_name libmain)
+  (modules (libmain))
+  (wrapped false)))
+
+(library
+ ((name libmain_main)
+  (public_name libmain.main)
+  (modules (libmain_main))
+  (libraries (libmain))
+  (wrapped false)))
```
[the remark this answers is not in the document]
Comment (rgrinberg) [reply libmain/jbuild]: Hmm, I usually recommend others to add `(wrapped false)` to 1 module libraries. The reason is that adding a module while forgetting to add `(wrapped false)` is a bit unintended.

OK, I'll remove the wrapping nonetheless.

**Comment by diml** [comment]
> I'm not sure what the right thing to do here, but I don't really like making artificial differences between benchmarks and tests, and I agree that the concept of a runner is murky to the user.

I suppose we could also add a field `inline_benchmarks`  for benchmarks specific parameters, both systems might need specific parameters that don't make sense for the other.

But before we decide on the details, I think there is a more important question to answer: should jbuilder have specific knowledge about ppx_inline_test, ppx_expect and ppx_bench?

Initially I was thinking no, and that's why I proposed the method implemented in this PR, which tries to abstract the interaction between the build system and the ppx systems. I did the same for ppx drivers and retrospectively I'm not sure that this was the right choice:
- it makes things more complicated when we want to add new features
- it's not clear that it's necessary given that there are currently no alternatives to the omp driver and ppx_driver and it's unlikely they'll ever be

And I feel that things will be even more complicated for ppx_inline_test, ppx_expect and ppx_bench given that the way they interact with the build system is more complex.

So I'm leaning towards replacing the generic ppx runner mechanism implemented in this PR by specific support for ppx_inline_test, ppx_expect and ppx_bench. That will get us support for inline tests and benchmarks quicker. If in the future there is a demand for alternatives, we could re-evaluate and try to generalize the system at this point.

**Comment by rgrinberg** [comment]
OK, I agree with your conclusion.

> it makes things more complicated when we want to add new features

Do you have some specific features in mind?

So basically we need to:

* Get rid of libmain
* Get rid of the changes to META files
* Hard code the last linked library for ppx_test/bench/expect etc.

Anything else?

File: libmain/jbuild (hunk 1 above)
[the remark this answers is not in the document]
Comment (diml) [reply libmain/jbuild]: Hmm, I see. Personally I would rather discourage the use of `wrapped false` in new code

**Comment by diml** [comment]
That seems right.

Later, we'll need to implement the partition stuff, which allows to run tests between different ml files in parallel. But we can do that in another PR.

**Comment by diml** [comment]
For ppx_expect, the runner needs to be linked with `ppx_expect.evaluator`

**Comment by rgrinberg** [comment]
In that case do we also need to link in `ppx_inline_test.runner`?

Also, one thing that I'm a bit confused by. If we have both inline tests and expect tests, do we expect to run them all in 1 binary? Or should we produce 2 binaries?

**Comment by diml** [comment]
> In that case do we also need to link in ppx_inline_test.runner?

yes

> Also, one thing that I'm a bit confused by. If we have both inline tests and expect tests, do we expect to run them all in 1 binary? Or should we produce 2 binaries?

1 binary.

Basically, `ppx_expect.evaluator` registers itself as a test evaluator against the `ppx_inline_test` runtime: https://github.com/janestreet/ppx_expect/blob/master/evaluator/ppx_expect_evaluator.ml#L123

**Comment by rgrinberg** [comment]
@diml ok I've removed all the extraneous stuff. The inline tests don't seem to be running because the `am_testing` variable doesn't seem to be set by the linked in stubs. I can't quite tell what the problem here is though. I am linking `ppx_inline_test.runner` last as necessary.

I still need to add an inline_tests stanza. That should also setup the expect tests as well, right?

File: src/inline_lib.ml [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+open Jbuild
+open Build.O
+open! No_io
+
+type rule =
+  { exe: Jbuild.Executables.t
+  ; alias_name: string
+  ; alias_action: (unit, Action.t) Build.t
+  ; alias_stamp: Sexp.t
+  ; last_lib: string
```
Comment (rgrinberg) [inline src/inline_lib.ml]: It might make sense to just add this field to the specification of `Jbuild.Executables.t`.

**Comment by diml** [comment]
Thinking about this again, I think we should keep the `inline_tests` field rather than change it to a stanza; since we will setup tests even when there is no `inline_tests` written in the jbuild file, it's better to have it as a field I think, which will just have a default value when not present. Otherwise it's like implicitly adding a stanza, which I'm not completely sure about.

**Comment by rgrinberg** [comment]
The issue that you've pointed with `last_lib` and the transitive closure is fixed.

We were discussing the issue with switching back to generating a source module to run the inline tests. The disadvantage of this approach of course is that this dummy source module can be visible to the user. We can workaround this by generating it in its own dir (like we do for utop), but then that changes from which directory the tests will be executed. I.e. `cd _build/default/src && runner.exe` vs. `cd _build/default/src/.inline_tests/runner.exe`. Probably not a big deal, but it does make such tests a bit different from manually defined tests.

File: src/gen_rules.ml [hunk 3]
```diff
@@ -519,13 +546,36 @@ module Gen(P : Params) = struct
...
+        | None ->
+          Build.fail { fail = fun () ->
+            die "@{<error>Error@}: I couldn't find '%s'.\n\
+                 Hint: opam install %s" lib lib
+          }
+          >>>
+          requires
+        | Some lib ->
+          requires >>^ (fun libs -> (List.filter ~f:(fun lib' ->
+            Lib.best_name lib <> Lib.best_name lib'
```
Comment (rgrinberg) [inline src/gen_rules.ml]: This is gross, but we're discussing switching things to source files anyway. So I will fix this only if we keep `last_lib`.

**Comment by diml** [comment]
What about naming the generated file `.something.ml-gen`?

**Comment by rgrinberg** [comment]
ok @diml I switched the code to generate a source based runner. Making the filename start with a `.` will take a bit more work to first refactor to `build_exe`. Let's address that later along with the utop code since it isn't critical.

**Comment by diml** [comment]
What about using `_` instead of `.`, would it be simpler? That would give us the same benefit.

> I've also added some hard coded support for setting up the tests when ppx_jane is used. My only real reservation is situations where a user might want to opt out of such tests but is still using the preprocessor using ppx_jane. I can't really think of when that's desirable, so I guess it's OK.

Hmm, that's unfortunate :/. It's not only ppx_jane, there is also ppx_jane_kernel and users might define their own aliases. What we need to do is scan the transitive closure rather than what's written in the jbuild file.

You can check this commit: de0f65b4f4cf0cf353a149d231e262179c3fe2e2. It removed knowledge of hard-coded ppx rewriters. We should restore some of it. What the code was doing before is that along with the generated `ppx.exe` in `_build/default/.ppx/ppx_blah+ppx_foo/`, it also generated an `info.sexp` file with the list of well known ppx rewriters that were  linked in the ppx.

**Comment by rgrinberg** [comment]
> What about using _ instead of ., would it be simpler? That would give us the same benefit.

I don't think it helps. I get an exception like this either way:

```
Description: "Mylib_test_runner" not found in map <dependency graph in _build/default>
Backtrace:
Raised at file "src/import.ml", line 19, characters 43-65
Called from file "src/top_closure.ml", line 24, characters 24-44
Called from file "src/top_closure.ml", line 33, characters 14-42
Called from file "src/top_closure.ml", line 37, characters 10-58
Called from file "src/ocamldep.ml", line 94, characters 8-54
Called from file "src/ocamldep.ml", line 102, characters 11-45
Called from file "src/build_system.ml", line 273, characters 16-33
Called from file "src/build_system.ml", line 259, characters 8-25
Called from file "src/build_system.ml", line 272, characters 16-33
Called from file "src/build_system.ml", line 259, characters 8-25
Called from file "src/build_system.ml", line 259, characters 8-25
Called from file "src/build_system.ml", line 259, characters 8-25
Called from file "src/build_system.ml", line 306, characters 17-47
Called from file "src/future.ml", line 88, characters 40-45
Called from file "src/future.ml", line 83, characters 65-70
Called from file "src/future.ml", line 37, characters 9-12
```

Ok I will look into looking up the deps in transitive closure.

**Comment by diml** [comment]
> Why is it necessary to setup the tests automatically?

It's not necessary but it's nicer for users. For instance, when using a ppx rewriter we don't have to specify the runtime dependencies manually since the information is already available to jbuilder. It should be the same for inline tests.

> Users of ppx_jane will not get inline tests automatically. Will save them some building time.

It should be small though, and it's only for running tests. Additionally, I think in general people are careful about their dependencies and will use ppx_expect only if they need it.

> It side steps the ppx info detection issue above. Which means that we won't have to set executable/alias rules based on the transitive deps of another lib.. I think that's pretty nice. Otherwise, I'm a bit stuck trying to generate a rule for an .exe conditionally.

What you can do is setup the rule for the .exe systematically and only setup running it if ppx_... is detected. When we have support for dynamic rules we'll be able to setup the rule for the .exe conditionally.

**Comment by hhugo** [comment]
What's the status on this one ? any ETA ?

**Comment by rgrinberg** [comment]
What remains is to finish off #479, implement a dependency closure that works outside the build arrow, and then re-implement this PR on top of that. For ETA, a week or two?

**Comment by rgrinberg** [comment]
> You can check this commit: de0f65b. It removed knowledge of hard-coded ppx rewriters. We should restore some of it. What the code was doing before is that along with the generated ppx.exe in _build/default/.ppx/ppx_blah+ppx_foo/, it also generated an info.sexp file with the list of well known ppx rewriters that were linked in the ppx.

@diml do we still need this?

If we can calculate the transitive closure outside the build arrow, why do we need the build system to dump this information? The information is only acessible from within the arrow anyway, so it's not accessible when generating executables_rules.  I think that all I need is prefix_rules and and a regular transitive closure to implement this.

**Comment by rgrinberg** [comment]
Here's another issue with auto-detection that's a bit tough to address. Suppose that the user is setting up inline tests using ppx_jane but ppx_jane is absent. Then `$ jbuilder runtest` will likely be a no-op if only inline tests were setup. This is because can't know if ppx_jane depends on ppx_inline_test without knowing that information. This case is not so dangerous, because the user will still fail to build the library. But I think it's a small advantage for manually enabling inline tests for a library.

**Comment by diml** [comment]
> @diml do we still need this?

No, we don't indeed.

> Here's another issue with auto-detection that's a bit tough to address. Suppose that the user is setting up inline tests using ppx_jane but ppx_jane is absent. Then $ jbuilder runtest will likely be a no-op if only inline tests were setup. This is because can't know if ppx_jane depends on ppx_inline_test without knowing that information. This case is not so dangerous, because the user will still fail to build the library. But I think it's a small advantage for manually enabling inline tests for a library.

Ah, that's true, I hadn't thought about that. Another solution is to scan the .ml files for `let%test`, ... That's what we do in our jenga rules, but I'm not excited at the idea of doing that in Dune.

I guess we can require tests to be enabled manually for now, and when we switch to Dune, we'll add a plugin `ppx_inline_test` which will cause libraries to be scanned this way. What do you think?

**Comment by yminsky** [comment]
That seems like a great compromise to me.

**Comment by diml** [comment]
Additionally, I asked a few people at lunch time and it seems that not everybody like the implicit behavior. Some people would prefer to just get an error when tests are present but no `(inline_tests ...)` field is present in the jbuild file.

Let's start with the explicit version then. That mean that we don't need to wait for the library management refactoring.

In parallel, I'll adapt ppx_inline_test so that we can implement the check. I think I'll add a flag that makes ppx_inline_test dump a file with a list of partitions. That will avoid the extra call to the test runner to get the list of partitions and will we can fail with the list of partitions is non-empty by no `(inline_tests ...)` field is present.

**Comment by rgrinberg** [comment]
> I guess we can require tests to be enabled manually for now, and when we switch to Dune, we'll add a plugin ppx_inline_test which will cause libraries to be scanned this way. What do you think?

Should we also move inline_test_deps into this `inline_tests` field as well? Do we need to think about how this will generalize to benchmarks? Just to make sure the syntax is future proof for adding support for those.

> In parallel, I'll adapt ppx_inline_test so that we can implement the check

I see. So basically something like a cookie arg to pass to the preprocessors `--cookie user_expects_inline_tests=false` which ppx_inline_test would detect and error out if tests are present?

**Comment by diml** [comment]
> Should we also move inline_test_deps into this inline_tests field as well?

Yes, I suggest:

```ocaml
(library
 ((name foo)
  ...
  (inline_tests ((deps (file.txt ...))
                       (flags (...))))))
```

(this is what we have in Jane Street jenga rules). `flags` is an additional list of flags to pass to the test runner.

> Do we need to think about how this will generalize to benchmarks? Just to make sure the syntax is future proof for adding support for those.

I suppose it will be an `inline_benchmarks` field.

> I see. So basically something like a cookie arg to pass to the preprocessors --cookie user_expects_inline_tests=false which ppx_inline_test would detect and error out if tests are present?

That's one way, or just `-no-inline-tests-expected` that we pass only when `ppx_inline_test` is used.

I was thinking of a flag `-dump-inline-tests-partitions <file>` where ppx inline tests would dump the list of partitions it found in the file. The file would be empty if no tests are defined. If the user didn't put an `inline_tests` field and `ppx_inline_test` is used, then we add an action in the `runtest` alias that checks that all the tests partition files are empty.

**Comment by rgrinberg** [comment]
One thing that's a bit annoying is that to specify inline_tests flag without an argument you gotta do `(inline_tests ())` rather than `(inline_tests)`. Should we allow the latter instead?

**Comment by diml** [comment]
Allowing `(inline_tests)` seems fine to me.

**Comment by rgrinberg** [comment]
OK, I've allowed the short syntax as well.

**Comment by rgrinberg** [comment]
@diml the ppx_expect tests are broken, is this expected?

Also, do we still need prefix_rules? We only use the transitive closure to detect ppx_expect, but it will already be recorded by Record_lib_deps when the rules are generated for the inline test running executable. And we catch the errors from calculating the transitive closure so that shouldn't be an issue either.

**Comment by diml** [comment]
> @diml the ppx_expect tests are broken, is this expected?

They need the development version of ppx_expect and ppx_inline_test to work. Once it's all working correctly, I'll backport the changes and release them in opam.

> Also, do we still need prefix_rules? We only use the transitive closure to detect ppx_expect, but it will already be recorded by Record_lib_deps when the rules are generated for the inline test running executable. And we catch the errors from calculating the transitive closure so that shouldn't be an issue either.

Sorry, I'm a bit lost here. `prefix_rules` is for when we'll change the transitive closure to be computed outside in the arrow. #516 will do that. Since we now require an additional field in library stanzas to setup and run tests, it's likely that we don't need to wait for this feature.

**Comment by rgrinberg** [comment]
> Sorry, I'm a bit lost here. prefix_rules is for when we'll change the transitive closure to be computed outside in the arrow. #516 will do that. Since we now require an additional field in library stanzas to setup and run tests, it's likely that we don't need to wait for this feature.

Gotcha. Is there anything else to do this for this PR apart from documentation then?

**Comment by diml** [comment]
> Gotcha. Is there anything else to do this for this PR apart from documentation then?

Documentation and testing that the full workflow for expect tests work as expected. I can do the latter at the same time as I backport the changes.

I'll try to review this PR today or tomorrow.

**Comment by diml** [comment]
I pushed support for expect tests. I also renamed `Inline_lib` to `Inline_tests`, it seemed clearer to me

**Comment by diml** [comment]
BTW, I'm using the v0.10 branches of ppx_inline_test and ppx_expect to test

**Comment by diml** [comment]
Ok, I think I'm done. @rgrinberg, you can review my last changes.

@c-cube, we have made the system configurable, so you should be able to make it work for qtest. Basically you have to write something like this in qtest itself:

```
(library
  ((name qtest)
   ...
   (inline_tests.backend
      ((runner_libraries (qtest))
       (gen (qtest-generator ${impl-files})) ;; How to generate the test code, ${impl-files} is the list of ml files
       (flags ( ... )) ;; flags for running the tests
       ))))
```

and in a library that contains tests:

```
(library
  ((name foo)
   ...
   (inline_tests ((backend qtest)))))
```

you can also simply write `(inline_tests)` if `qtest` is a dependency of `foo`.

**Comment by c-cube** [comment]
:heart_eyes:

**Comment by c-cube** [comment]
Just a note: qtest is really only a generator (so the "qtest-generator" part is qtest itself) and doesn't have a library. However, the files it generates depend on ounit and qcheck , which are separate projects.

**Comment by vyorkin** [comment]
> Add support for ppx_bench

Is there any working example of how to use `ppx_bench`?
I've found this issue, but still have no idea of how to use it https://github.com/janestreet/core_bench/issues/12#issuecomment-532624589

**Comment by diml** [comment]
There is one in dune itself, in the `bench` directory

**Comment by vyorkin** [comment]
Thanks! not sure what I'm doing wrong, I've even tried to make my example look exactly the same as in `dune/bench`: https://github.com/vyorkin/core_profiler_playground/tree/master/bench -- still getting the `No benchmarks to run!` output

**Comment by diml** [comment]
Could you try adding `(link_flags -linkall)` to your executable stanza?

Otherwise yeah, the runner.sh bit is a bit annoying. Basically, we'd like to refactor a bit core_bench so that such a runner wrapper is not needed, and then we'll be able to add first class support for inline benchmarks in dune, just like for inline tests.

**Comment by vyorkin** [comment]
> Could you try adding (link_flags -linkall) to your executable stanza?

Thanks, got it working!
